### ISSUE TITLE
Allowing a dynamic email subject

### ISSUE THREAD

**Issue by hanne** [mike-charles/python-buffering-smtp-handler#1]
Hi,

Thanks for posting this, it certainly saved me some time!

I needed the email subject to be a bit more dynamic, so I've modified the code to allow the subject input to be a function which can take the message body text as input and return a string with the resulting subject line. In my case, this function will then count the number of occurrences of certain strings in the message body, and report the numbers in the email subject.

I hope it's useful.

    def flush(self):
        if len(self.buffer) > 0:
            try:
                import smtplib
                port = self.mailport
                if not port:
                    port = smtplib.SMTP_PORT
                smtp = smtplib.SMTP(self.mailhost, port)
                if isinstance(self.toaddrs, list):  # If toaddrs is a list, then join them as a string
                    toaddrs = ','.join(self.toaddrs)
                else:
                    toaddrs = self.toaddrs
                msg_body = ""
                for record in self.buffer:
                    s = self.format(record)
                    msg_body += s + "\r\n"
                if callable(self.subject):
                    subject = self.subject(msg_body)
                else:
                    subject = self.subject
                msg = "From: {}\r\nTo: {}\r\nSubject: {}\r\n\r\n{}".format(self.fromaddr, toaddrs,
                                                                           subject, msg_body)
                smtp.sendmail(self.fromaddr, self.toaddrs, msg)
                smtp.quit()
            except:
                self.handleError(None)  # no particular record
            self.buffer = []

**Comment by mikecharles** [comment]
Thanks @hanne ! Can you also share a snippet of what your subject function looks like?

**Comment by hanne** [comment]
Sure, it'll be something like this:

    def summarize_subject(contents):
        err = "error"
        warn = "warning"
        e_cnt = 0
        w_cnt = 0
        # Count the number of lines containing the flags
        for line in contents.split("\n"):
            if err in line:
                e_cnt += 1
            if warn in line:
                w_cnt += 1
        return "Output from Data Monitor - {} errors, {} warnings".format(e_cnt, w_cnt)

It's used with a scheduled nightly task, so it's convenient to have the summary in the subject line. And then I've set up my email to automatically mark it as read if it contains " 0 errors".

**Comment by mikecharles** [comment]
Very cool. Are you interested in making a Pull Request with this update? Seems handy, and it seems like it falls back to using the specified subject as a string if it's not a method, correct?

**Comment by hanne** [comment]
Yes, it still works the same way as before, if it's not a method.
If you want, you're very welcome to just take the code and include it.

**Comment by mikecharles** [comment]
Ok thanks, I'll keep this issue until then
